devito/arguments.py

```diff
@@ -158,198 +145,398 @@ class PtrArgument(Argument):
+        arg = ScalarArgument(o.name, [dependency], dtype=o.dtype)
+        arg.provider = o
+        return arg
 
-class ConstantArgProvider(ArgumentProvider):
 
-    """ Class used to decorate Constant Data objects with behaviour required for runtime
-        arguments.
+class ValueVisitor(GenericVisitor):
+    """Visits types to derive their value
     """
+    def __init__(self, consumer, known_values):
+        self.consumer = consumer
+        self.known_values = dict([(k, v) for k, v in known_values.items()
+                                  if v is not None])
+        super(ValueVisitor, self).__init__()
+
+    def visit_Function(self, o, param=None):
+        assert(isinstance(self.consumer, TensorArgument))
+        try:
+            return o._data_buffer
+        except AttributeError:
+            return o.data
 
-    @cached_property
-    def rtargs(self):
-        return (ScalarArgument(self.name, self, default_value=self.data),)
+    def visit_Constant(self, o, param=None):
+        return o.data
 
+    def visit_Dependency(self, o):
+        return self.visit(o.obj, o.param)
 
-class TensorFunctionArgProvider(ArgumentProvider):
+    def visit_function(self, o, param=None):
+        return o(self.consumer, self.known_values, param)
 
-    """ Class used to decorate Symbolic Data objects with behaviour required for runtime
-        arguments.
-    """
+    def visit_Object(self, o, param=None):
+        if callable(o.value):
+            return o.value()
+        else:
+            return o.value
 
-    @cached_property
-    def rtargs(self):
-        return (TensorArgument(self.name, self),)
+    def visit_object(self, o, param=None):
+        return o
 
+    def visit_DimensionParameter(self, o, param=None):
+        """ Called when some Argument's value is being derived and the dependency tree
+            was followed till a DimensionParameter object. The Argument for which we are
+            deriving the value is self.consumer
+        """
 
-class ScalarArgProvider(ArgumentProvider):
+        # We are being asked to provide a default value for dim_start
+        if self.consumer.name == o.provider.start_name:
+            return 0
 
-    """ Class used to decorate Scalar Function objects with behaviour required for runtime
-        arguments.
+        if o in self.known_values and not isinstance(self.known_values[o],
+                                                     UnevaluatedDependency):
+            provided_values = [self.known_values[o]]
+        else:
+            provided_values = [get_value(o, x, self.known_values)
+                               for x in o.gets_value_from]
+
+        if len(provided_values) > 0:
+            if not all(x is not None for x in provided_values):
+                unknown_args = [x.obj for x in o.gets_value_from
+                                if get_value(o, x, self.known_values) is None]
+
+                def late_evaluate_dim_size(consumer, known_values, partial_values):
```

Comment:
when/why is this used/needed?

Reply:
If an argument depends on the size of an autotuned dimension, the argument's value can not be determined at the time this derivation routine is run. This lazy evaluation routine is called to derive the value of the arguments once the size of the dimension has been derived by the autotuner.